graphs.hcu

```diff
@@ -0,0 +1,208 @@
+#include "util.hpp"
+#include "util_cuda.h"
+#include "kernels.hcu"
+
+cudaGraphNode_t add_empty_node(cudaGraph_t& graph) {
+    cudaGraphNode_t node = {0};
+    cuda_api(cudaGraphAddEmptyNode, &node, graph, nullptr, 0);
+    return node;
+}
+
+template<class K>
+cudaGraphNode_t add_kernel_node(cudaGraph_t& graph, const benchmark_parameters& p, K kernel, const std::vector<void*> args) {
+    cudaGraphNode_t node = {0};
+    cudaKernelNodeParams params = {0};
+    params.func           = (void*) kernel;
+    params.gridDim        = p.blocks;
+    params.blockDim       = p.threads;
+    params.sharedMemBytes = 0;
+    params.kernelParams   = (void**) args.data();
+    params.extra          = nullptr;
+
+    cuda_api(cudaGraphAddKernelNode, &node, graph, nullptr, 0, &params);
+
+    return node;
+}
+
+template<int N, int ix, typename... Ts>
+constexpr void set_args(std::unique_ptr<std::tuple<Ts...>>& values, std::vector<void*>& pointers) {
+    if constexpr(ix >= N) {
+        return;
+    } else {
+        pointers[ix] = &std::get<ix>(*values);
+        set_args<N, ix + 1, Ts...>(values, pointers);
+    }
+}
+
+void add_dependencies(cudaGraph_t& graph, const cudaGraphNode_t& from, const cudaGraphNode_t& to) {
+    cuda_api(cudaGraphAddDependencies, graph, &from, &to, 1);
+}
+
+void add_dependencies(cudaGraph_t& graph, const cudaGraphNode_t& from_, const std::vector<cudaGraphNode_t>& to) {
+    auto n = to.size();
+    std::vector<cudaGraphNode_t> from(n, from_);
+    cuda_api(cudaGraphAddDependencies, graph, from.data(), to.data(), n);
+}
+
+void add_dependencies(cudaGraph_t& graph, const std::vector<cudaGraphNode_t>& from, const cudaGraphNode_t& to_) {
+    auto n = from.size();
+    std::vector<cudaGraphNode_t> to(n, to_);
+    cuda_api(cudaGraphAddDependencies, graph, from.data(), to.data(), n);
+}
+
+void add_dependencies(cudaGraph_t& graph, const std::vector<cudaGraphNode_t>& from, const std::vector<cudaGraphNode_t>& to) {
+    assert(to.size() == from.size());
+    cuda_api(cudaGraphAddDependencies, graph, from.data(), to.data(), to.size());
+}
+
+
+template<typename K, typename... As>
+auto make_graph(const benchmark_parameters& p, K func, As... as) {
+    std::vector<void*> args(sizeof...(As), nullptr);
+    auto tmp = std::make_unique<std::tuple<As...>>(as...);
+    set_args<sizeof...(As), 0, As...>(tmp, args);
```

Comment:
Is this way of doing things necessary? I'd rather keep things easy to understand.

Reply:
Yes, as far as I understand. The issues are
- kernel_node_params takes an array of void*
- that means all arguments need their address to be taken
- consequently these pointers must be valid throughout the graph lifetime
- that includes pass-by-value parameters

Reply:
Okay then can you please annotated the code with some comments?

Reply:
I re-read the docs. It seems that the data is copied, so I removed most of the crazyness. Tests pending, though.